src/clientRPC/ClientClient.ts

```diff
@@ -0,0 +1,345 @@
+    const connectProm = promise<void>();
+    const authenticateProm = promise<NodeId>();
+    const ws = new WebSocket(address, {
+      rejectUnauthorized: false,
+    });
+    // Creating logic for awaiting active connections and terminating them
+    const abortHandler = () => {
+      ws.terminate();
+    };
+    const abortController = new AbortController();
+    const activeConnectionProm = new PromiseCancellable<void>((resolve) => {
+      ws.once('close', () => {
+        abortController.signal.removeEventListener('abort', abortHandler);
+        resolve();
+      });
+    }, abortController);
+    abortController.signal.addEventListener('abort', abortHandler);
+    this.activeConnections.add(activeConnectionProm);
+    activeConnectionProm.finally(() =>
+      this.activeConnections.delete(activeConnectionProm),
+    );
+    // Handle connection failure
+    const openErrorHandler = (e) => {
+      connectProm.rejectP(
+        new clientRPCErrors.ErrorClientConnectionFailed(undefined, {
+          cause: e,
+        }),
+      );
+    };
+    ws.once('error', openErrorHandler);
+    // Authenticate server's certificates
+    ws.once('upgrade', async (request) => {
+      const tlsSocket = request.socket as TLSSocket;
+      const peerCert = tlsSocket.getPeerCertificate(true);
+      clientRpcUtils
+        .verifyServerCertificateChain(
+          this.expectedNodeIds,
+          clientRpcUtils.detailedToCertChain(peerCert),
+        )
+        .then(authenticateProm.resolveP, authenticateProm.rejectP);
+    });
+    ws.once('open', () => {
+      this.logger.info('starting connection');
+      connectProm.resolveP();
+    });
+    // TODO: Race with a connection timeout here
+    // There are 3 resolve conditions here.
+    //  1. Connection established and authenticated
+    //  2. connection error or authentication failure
+    //  3. connection timed out
+    try {
+      const result = await Promise.race([
+        timer?.then(() => timeoutSymbol) ?? new Promise(() => {}),
+        await Promise.all([authenticateProm.p, connectProm.p]),
+      ]);
+      if (result === timeoutSymbol) {
+        throw new clientRPCErrors.ErrorClientConnectionTimedOut();
+      }
+    } catch (e) {
+      // Clean up
+      // unregister handlers
+      ws.removeAllListeners('error');
+      ws.removeAllListeners('upgrade');
+      ws.removeAllListeners('open');
+      // Close the ws if it's open at this stage
+      ws.terminate();
+      // Ensure the connection is removed from the active connection set before
+      //  returning.
+      await activeConnectionProm;
+      throw e;
+    }
+    // Cleaning up connection error
+    ws.removeEventListener('error', openErrorHandler);
+
+    let readableClosed = false;
+    let writableClosed = false;
+    const readableLogger = this.logger.getChild('readable');
+    const writableLogger = this.logger.getChild('writable');
+    const readableStream = new ReadableStream<Uint8Array>(
```

Comment:
This portion along with shared with WSS, can be extracted out to its own `WebSocketConnection` class that encapsulates the logic of `ReadableStream` and `WritableStream` similar to what I'm doing with `QUICConnection` and `QUICStream`.

Reply:
Unfortunately the client and server implementation of the streams are different enough that they can't be shared. But I can have them share an interface or an abstract class so they can implement the same behaviour. 

As for the features the'll have, we'd need...
1. Extend the `ReadableWritablePair<Uint8Array, Uint8Array>` object.
2. Flag for when each and or both streams have ended.
3. Promises that resolve when each and or both streams have ended and reject with errors.
4. Ability to abort the active connection

This class can be set into the active map for both the client and server.

Reply:
What would be so different between the 2?

In the quic system, it's the same except how it gets constructed/created.

That on the client side, created by connecting. Whereas on server side, created by accepting.

Reply:
I've created an abstract class `WebSocketStream` that standardises tracking the state of the readable, writable and websocket. It also provides an end method that needs to be implemented. It implements the `ReadableWritablePair` and can be used as such. 

It standardises the interface between the client stream and server stream. Bot the client and server extend this for their internal stream representation. The `ActiveConnections` set now contains these objects.

We can now add on to the abstract class when we need to enforce any more features for the streams.

Reply:
I see. Well you should still separate into the classes for the connection. Yes you can then share an interface or an abstract class if they have shared implementation.

Reply:
This is done.